### PR TITLE
feat: support compiling ln-dlc-node under wasm

### LINKED ISSUE #2167: Replace `thread::spawn` and `spawn_blocking` in `ln-dlc-node`

### LINKED ISSUE #2209: Make ln-dlc-node compile under WASM

### PR DESCRIPTION
**Pull request by Restioson**
This PR allows `ln-dlc-node` to be compiled under wasm(32-unknown-unknown) :tada: (when `ln_net_tcp is disabled and `ln_net_ws` is enabled)

This is a relatively big change in that I've async-ified a lot of the interfaces here, which is why I've requested reviews from everyone. Another alternative is to use more of `block_on`

Fixes #2209, fixes #2167

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -33,9 +33,8 @@ lightning-background-processor = { git = "https://github.com/bonomat/rust-lightn
 lightning-transaction-sync = { git = "https://github.com/bonomat/rust-lightning-p2p-derivatives", rev = "e49030e" }
 lightning-persister = { git = "https://github.com/bonomat/rust-lightning-p2p-derivatives", rev = "e49030e" }
 
-rust-bitcoin-coin-selection = { git = "https://github.com/p2pderivatives/rust-bitcoin-coin-selection" }
-
-esplora-client = { git = "https://github.com/bitcoindevkit/rust-esplora-client", rev = "269360f" }
+rust-bitcoin-coin-selection = { git = "https://github.com/p2pderivatives/rust-bitcoin-coin-selection", rev = "4054519" }
```
Comment (luckysori) [inline Cargo.toml:36]: ❓ Is this even used?

File: DEVELOPMENT.md [hunk 2]
```diff
@@ -79,7 +79,7 @@ For more information on what containers are available please have a look at the
 To start the development environment you can just run:
 
 ```bash
-docker-compose up
+docker compose up
```
Comment (luckysori) [inline DEVELOPMENT.md:82]: 🔧 Let's put this change in a separate patch, please.

File: coordinator/src/collaborative_revert.rs [hunk 3]
```diff
@@ -227,8 +228,7 @@ pub fn confirm_collaborative_revert(
         "Broadcasting collaborative revert transaction"
     );
 
-    node.blockchain
-        .broadcast_transaction_blocking(&revert_transaction)
+    block_on(node.blockchain.broadcast_transaction(&revert_transaction))
```
Comment (luckysori) [inline coordinator/src/collaborative_revert.rs:231]: ❓ What runtime are we spawning this on?

File: crates/ln-dlc-node/src/blockchain.rs [hunk 4]
```diff
@@ -20,91 +21,69 @@ const SOCKET_TIMEOUT: u64 = 30;
 #[derive(Clone)]
 pub struct Blockchain<N> {
     /// Async client used during on-chain syncing and, sometimes, to broadcast transactions.
-    pub(crate) esplora_client_async: esplora_client::AsyncClient,
-    /// Blocking client used when the task to be performed is in a blocking context (usually
-    /// blocking trait methods).
-    esplora_client_blocking: esplora_client::BlockingClient,
```
Comment (luckysori) [inline crates/ln-dlc-node/src/blockchain.rs:26]: ♥️

File: crates/ln-dlc-node/src/dlc_wallet.rs [hunk 5]
```diff
@@ -58,16 +59,15 @@ impl<D, S, N> DlcWallet<D, S, N> {
...
 {
     fn send_transaction(
         &self,
         tx: &bitcoin_old::Transaction,
     ) -> Result<(), dlc_manager::error::Error> {
         let tx = to_tx_30(tx.clone());
 
-        self.blockchain
-            .broadcast_transaction_blocking(&tx)
+        block_on(self.blockchain.broadcast_transaction(&tx))
```
Comment (luckysori) [inline crates/ln-dlc-node/src/dlc_wallet.rs:70]: ❓ Don't we want to use the `tokio` runtime instead?

File: crates/ln-dlc-node/src/ln/app_event_handler.rs [hunk 6]
```diff
@@ -7,24 +7,24 @@ use crate::storage::TenTenOneStorage;
...
-impl<D: BdkStorage, S: TenTenOneStorage, N: Storage + Sync + Send> AppEventHandler<D, S, N> {
+impl<D: BdkStorage, S: TenTenOneStorage, N: Storage + Sync + Send + 'static>
+    AppEventHandler<D, S, N>
+{
     pub fn new(node: Arc<Node<D, S, N>>, event_sender: Option<EventSender>) -> Self {
         Self { node, event_sender }
     }
 }
 
-#[async_trait]
```
Comment (luckysori) [inline crates/ln-dlc-node/src/ln/app_event_handler.rs:27]: ♥️

File: crates/ln-dlc-node/src/ln/event_handler.rs [hunk 7]
```diff
@@ -1,29 +1,46 @@
...
+impl<T, F> EventHandlerFuture<T> for F where F: Future<Output = T> + Send + Sync {}
+
+// Under WASM, we spawn tasks using wasm_bindgen_futures, so Send + Sync is not required.
+// Additionally, they _cannot_ be Send + Sync, as the esplora-client futures are not Send + Sync
+// since they bind to JavaScript promises (which are not Send + Sync)
+#[cfg(target_arch = "wasm32")]
+pub trait EventHandlerFuture<T>: Future<Output = T> {}
+
+#[cfg(target_arch = "wasm32")]
+impl<T, F> EventHandlerFuture<T> for F where F: Future<Output = T> {}
```
Comment (luckysori) [inline crates/ln-dlc-node/src/ln/event_handler.rs:23]: Luckily we're getting rid of this soon, so this complexity will go away.

File: crates/ln-dlc-node/src/node/mod.rs [hunk 8]
```diff
@@ -80,6 +80,7 @@ pub mod dlc_channel;
 pub mod event;
 pub mod peer_manager;
 
+use crate::spawn;
```
Comment (luckysori) [inline crates/ln-dlc-node/src/node/mod.rs:83]: 🔧 `use` statements at the top, please.

File: crates/ln-dlc-node/src/lib.rs [hunk 9]
```diff
@@ -48,6 +48,7 @@ pub use ln::EventSender;
 pub use on_chain_wallet::ConfirmationStatus;
 pub use on_chain_wallet::EstimateFeeError;
 pub use on_chain_wallet::TransactionDetails;
+pub use tokio_with_wasm::tokio::spawn;
```
Comment (luckysori) [inline crates/ln-dlc-node/src/lib.rs:51]: I suppose we will need to start using this in `native` too. My understanding this that there will be no downsides for the `coordinator`, given that we won't be compiling for `wasm` there.

Comment (luckysori) [reply crates/ln-dlc-node/src/lib.rs:51]: To avoid problems, we could emit a compiler warning if we use `tokio::spawn` in `ln-dlc-node` and eventually `native`. I think we did something like that in `itchysats`.

**Review by luckysori — COMMENTED** [review]
Looking good, although I'm not sure about the user of `block_on`.

**Comment by bonomat** [comment]
I think we can close this for the time being. Thank you for all the hard work @Restioson
